### PR TITLE
MNIST enhanced performance. Faster convergence, better score

### PR DESCRIPTION
**Pull request by pliptor**
This patch credit goes to @beru in my experiments with the activation function in the last layer of the network. This is just a one line patch in the training file for the MNIST example that does not touch the network configuration. It only changes the batch size from 10 to 16. 

Convergence is at least 10% faster in my AMD Quad-Core A6 with SSE and TBB enabled.
Final classification accuracy is 99.11%. Results attached.

It appears that AVX must be disabled in order to achieve the above classification  accuracy. Thanks.

[mninst-train-batch16.txt](https://github.com/tiny-dnn/tiny-dnn/files/769672/mninst-train-batch16.txt)

Added one more patch to address the issue of users testing larger mini-batch sizes #538 
The current training code for MNIST example scales the learning rate of the optimizer by the
square root of the mini-batch size. This negatively impacts the convergence if the batch size is large. The first patch in this pull changed the batch size to 16 with positive results. Hence 
the limiting factor has been changed to the minimum of sqrt(16)=4 and the square root of the mini-batch size. Thanks.

### CODE REVIEW

File: examples/mnist/train.cpp [hunk 1]
```diff
@@ -79,10 +79,11 @@ static void train_lenet(const std::string& data_dir_path) {
 
   progress_display disp(static_cast<unsigned long>(train_images.size()));
   timer t;
-  int minibatch_size = 10;
+  int minibatch_size = 16;
   int num_epochs     = 30;
 
-  optimizer.alpha *= static_cast<tiny_dnn::float_t>(std::sqrt(minibatch_size));
+  optimizer.alpha *= std::min(
+    float_t(4), static_cast<tiny_dnn::float_t>(std::sqrt(minibatch_size)));
```
Comment (Randl) [inline examples/mnist/train.cpp]: @edgarriba @bhack Any ideas for better solution? If not, LGTM

Comment (Randl) [inline examples/mnist/train.cpp]: `std::min(tiny_dnn::float_t(4), static_cast<tiny_dnn::float_t>(std::sqrt(minibatch_size)));`

Comment (pliptor) [inline examples/mnist/train.cpp]: Yep. What's the best way to fix the repo? Reset back one commit and recommit or add another one?

Comment (Randl) [inline examples/mnist/train.cpp]: Whatever is convenient for you. Since we're squashing PRs in one commit, there is no much difference for us

**Comment by pliptor** [comment]
@Randl Thanks for your patience. I tried to consolidate them for cleanness but they got split. I think I can still rebase it but it this is okay I'll leave it as is.

**Comment by pliptor** [comment]
Thank you @Randl

File: examples/mnist/train.cpp [hunk 2]
```diff
@@ -79,10 +79,12 @@ static void train_lenet(const std::string& data_dir_path) {
...
   progress_display disp(static_cast<unsigned long>(train_images.size()));
   timer t;
-  int minibatch_size = 10;
+  int minibatch_size = 16;
   int num_epochs     = 30;
 
-  optimizer.alpha *= static_cast<tiny_dnn::float_t>(std::sqrt(minibatch_size));
+  optimizer.alpha *=
+    std::min(tiny_dnn::float_t(4),
+             static_cast<tiny_dnn::float_t>(std::sqrt(minibatch_size)));
```
Comment (beru) [inline examples/mnist/train.cpp]: I'm not sure how this change is needed. Aside from that, `std::sqrt` function has some overloads.
http://en.cppreference.com/w/cpp/numeric/math/sqrt, so rather than casting returned value, casting int argument to `tiny_dnn::float_t` might be another way to suppress a warning.

Comment (Randl) [inline examples/mnist/train.cpp]: For bigger batches, learning rate becomes too high, so we need to limit it somehow

Comment (bhack) [inline examples/mnist/train.cpp]: For small batch  https://github.com/tensorflow/tensorflow/issues/7476

**Comment by edgarriba** [comment]
it's okay to have good examples showing good performance, but I don't think we should care about optimizing time with this ... as a library (like others) in training mode the learning rate policy is up to the user.

**Comment by edgarriba** [comment]
LGTM

**Comment by pliptor** [comment]
@edgarriba Thanks. You have a good point. I forked the library to play instead with optimization over there.  This library is however very awesome for a learning platform for those that want C++. I'm already getting better than 99.2% with half the complexity of the example by just playing with it for a couple of days.

**Comment by edgarriba** [comment]
@pliptor thx for the feedback. In case you get more improvements don't hesitate to report them. Besides, could you add short comments in source showing where this formula comes from, or just cite if you got it from a paper. At first glance, it looks a magic number.

File: examples/mnist/train.cpp (hunk 2 above)
Comment (pliptor) [inline examples/mnist/train.cpp]: @bhack The conclusions of the paper end with "We have demonstrated that Batch Normalization, while effective, is not well suited to small or non-i.i.d. training minibatches."

I tried to track down with blame what was the rationale for line 85 but maybe only the original author knows it. It seems mini-batch processing reduces the learning rate by O(sqrt(batch_size)) (I'll try to find the paper). I believe that's why that multiplying constant was added to it. I mentioned @Randl this fix is a hack to address a user that was having trouble with the code by playing with the mini-batch size.

**Comment by pliptor** [comment]
@edgarriba No, unfortunately it appears to me line 95 was a hack fix for a hack :) There is a paper that I did find trying to find the reason behind original line. I'll try to find it.

**Comment by bhack** [comment]
https://github.com/NVIDIA/DIGITS/issues/227

**Comment by pliptor** [comment]
@bhack No, that was not the paper I saw. The one I saw had analytical derivation. The one you are pointing out seem empirical. 

The empirical :) observation that I had while playing with the code is that it is clear that mini-bathing 
reduces convergence rate. So I found line 95 justified. But I find it strange that sqrt(mini-batch) is already bad for batches larger than 16 or so with this setup. multiplying by mini-batch would make it even worse. But note there is another important variable (or two) the optimization algorithm itself and the initial learning rate value. So it doesn't seem to me there is a simple answer here if you take these all into consideration.

**Comment by pliptor** [comment]
Here's the paper.

http://www.cs.cmu.edu/~muli/file/minibatch_sgd.pdf

One really weird thing in the tests was the jump at epoch 16 for batch size 64. So initially batch size 64 looked very bad but eventually it caught up with batch 16.

File: examples/mnist/train.cpp (hunk 2 above)
Comment (pliptor) [inline examples/mnist/train.cpp]: @beru In numbers, say we pick minibatch_size = 16. The default start learning rate set by tiny-dnn in adagrad is alpha=0.01. We end up with alpha = 0.04 with the HEAD and it the algorithm takes it well: convergence speeds up without stalling it. But alpha bigger than that is making the algorithm to stall. 

Regarding the casting, yes, either way suppresses the warning although the scaling value would be different. The way you suggest would make the alpha scaling more conservative because the sqrt would be floored.  Either way is fine with me. Let me know.

Comment (beru) [inline examples/mnist/train.cpp]: @Randl Thank you for the answer! :bow: I haven't studied about neural networks so there're a lot to learn. 

@pliptor Than you for the detailed explanation. :joy: It's pretty helpful. I'm not sure if maximum of alpha should be adjusted per network. Talking about the casting, I need to check its consequences.

Comment (pliptor) [inline examples/mnist/train.cpp]: @beru This summary looks nice http://sebastianruder.com/optimizing-gradient-descent/

I agree with @edgarriba  it's not worth spending too much time in the optimization for the sake of writing the library. Maybe the easiest way to close this subject would be to run two simulations with each of the casting and pick the one that gives better result?

**Comment by edgarriba** [comment]
@beru give green light so that I can merge

File: examples/mnist/train.cpp (hunk 2 above)
Comment (beru) [inline examples/mnist/train.cpp]: @pliptor Thank you for introducing the page. Actually, I searched about the optimization and found several informative pages such as these and also the one you introduced.
http://cs231n.github.io/optimization-1/
http://int8.io/comparison-of-optimization-techniques-stochastic-gradient-descent-momentum-adagrad-and-adadelta/

I wish I can understand basics of deep learning someday...

Comment (pliptor) [inline examples/mnist/train.cpp]: @beru  I like those two links you are referencing. Time to study them!
